File: ios/InputPilot/Services/DeviceRepository.swift
import Foundation
import SwiftData

enum DeviceRepositoryError: Error, LocalizedError {
    case deviceUnreachable
    case notFound
    case alreadyExists(displayName: String)
    case secureProtocolRequired

    var errorDescription: String? {
        switch self {
        case .deviceUnreachable:
            "Could not reach the device on the local network."
        case .notFound:
            "Device not found."
        case let .alreadyExists(displayName):
            SavedDeviceIndex.alreadyExistsMessage(displayName: displayName)
        case .secureProtocolRequired:
            "This device must be reflashed with Secure Protocol v2 firmware."
        }
    }
}

@MainActor
final class DeviceRepository {
    private let context: ModelContext

    init(context: ModelContext) {
        self.context = context
    }

    func refreshAll(devices: [StoredDevice], api: any DeviceAPIClientProtocol) async -> Set<String> {
        let snapshots = devices.map { device in
            (
                deviceId: device.deviceId,
                mdnsHost: device.mdnsHost,
                staIP: device.staIP
            )
        }

        let results = await withTaskGroup(of: (String, DeviceStatus?, String).self) { group in
            for snapshot in snapshots {
                group.addTask {
                    let urls = DeviceEndpointResolver.endpointURLs(
                        mdnsHost: snapshot.mdnsHost,
                        staIP: snapshot.staIP
                    )
                    for url in urls {
                        if let status = try? await api.status(baseURL: url) {
                            return (snapshot.deviceId, status, url.host ?? snapshot.mdnsHost)
                        }
                    }
                    return (snapshot.deviceId, nil, snapshot.mdnsHost)
                }
            }

            var collected: [(String, DeviceStatus?, String)] = []
            for await result in group {
                collected.append(result)
            }
            return collected
        }

        var failed = Set<String>()
        for (deviceId, status, fallbackHost) in results {
            guard let device = devices.first(where: { $0.deviceId == deviceId }) else { continue }
            if let status,
               status.deviceId?.caseInsensitiveCompare(deviceId) == .orderedSame,
               status.protocolVersion == 2,
               status.capabilities.contains("secure_protocol_v2") {
                applyStatus(status, to: device, fallbackHost: fallbackHost)
            } else {
                failed.insert(deviceId)
            }
        }

        try? context.save()
        return failed
    }

    @discardableResult
    func refresh(device: StoredDevice, api: any DeviceAPIClientProtocol) async -> Bool {
        await refreshAll(devices: [device], api: api).isEmpty
    }

    func addFromDiscovery(
        status: DeviceStatus,
        fallbackHost: String,
        displayName: String
    ) async throws -> StoredDevice {
        guard let deviceId = status.deviceId?.lowercased(),
              deviceId.count == 12, deviceId.allSatisfy(\.isHexDigit),
              status.protocolVersion == 2,
              status.capabilities.contains("secure_protocol_v2") else {
            throw DeviceRepositoryError.secureProtocolRequired
        }
        if let existing = try fetchStored(deviceId: deviceId) {
            DeviceMerge.wifi(status, fallbackHost: fallbackHost, into: existing)
            try context.save()
            return existing
        }
        var device = Device(status: status, deviceId: deviceId, endpointHost: fallbackHost)
        device.displayName = displayName
        try DeviceStore.upsert(device, in: context)
        guard let stored = try fetchStored(deviceId: deviceId) else {
            throw DeviceRepositoryError.notFound
        }
        return stored
    }

    func addOrMergeBluetooth(metadata: BLEDeviceMetadata, displayName: String) throws -> StoredDevice {
        guard metadata.deviceId.count == 12,
              metadata.deviceId.allSatisfy(\.isHexDigit),
              metadata.protocolVersion == 2,
              metadata.capabilities.contains("secure_protocol_v2") else {
            throw DeviceRepositoryError.secureProtocolRequired
        }
        if let existing = try fetchStored(deviceId: metadata.deviceId) {
            DeviceMerge.bluetooth(metadata, into: existing)
            try context.save()
            return existing
        }
        let stored = StoredDevice(deviceId: metadata.deviceId, displayName: displayName, mdnsHost: "", staIP: nil,
            lastSeen: Date(), firmwareVersion: metadata.firmware,
            protocolVersion: metadata.protocolVersion, capabilities: metadata.capabilities,
            lastCapabilitiesUpdate: Date(), otaSchema: metadata.otaSchema, bluetoothDiscovered: true)
        context.insert(stored)
        try context.save()
        return stored
    }
    func setJiggle(_ device: StoredDevice, enabled: Bool) async throws {
        let settings = KeepAwakeSettings(moveEnabled: enabled, moveIntervalMs: device.moveIntervalMs,
            clickEnabled: device.clickEnabled, clickIntervalMs: device.clickIntervalMs)
        try await setKeepAwake(device, settings: settings)
    }

    func setKeepAwake(
        _ device: StoredDevice,
        settings: KeepAwakeSettings
    ) async throws {
        let bluetooth = InputPilotBluetoothManager.session(deviceId: device.deviceId)
        do { try await bluetooth.setKeepAwake(settings) }
        catch {
            let host = device.staIP ?? device.mdnsHost
            guard !host.isEmpty else { throw error }
            let wifi = InputPilotWiFiManager.session(host: host, deviceId: device.deviceId)
            try await wifi.setKeepAwake(settings)
        }
        apply(settings, to: device); try context.save()
    }

    func apply(_ settings: KeepAwakeSettings, to device: StoredDevice) {
        device.jiggleEnabled = settings.moveEnabled
        device.moveIntervalMs = settings.moveIntervalMs
        device.clickEnabled = settings.clickEnabled
        device.clickIntervalMs = settings.clickIntervalMs
    }

    func rename(_ device: StoredDevice, name: String) throws {
        device.displayName = name
        try context.save()
    }

    func delete(_ device: StoredDevice) throws {
        PairingKeyStore.remove(deviceId: device.deviceId)
        Task { await InputPilotWiFiManager.removeSessions(deviceId: device.deviceId) }
        context.delete(device)
        try context.save()
    }

    private func applyStatus(_ status: DeviceStatus, to stored: StoredDevice, fallbackHost: String) {
        DeviceMerge.wifi(status, fallbackHost: fallbackHost, into: stored)
    }

    private func fetchAll() throws -> [StoredDevice] {
        try context.fetch(FetchDescriptor<StoredDevice>())
    }

    private func fetchStored(deviceId: String) throws -> StoredDevice? {
        var descriptor = FetchDescriptor<StoredDevice>(
            predicate: #Predicate { $0.deviceId == deviceId }
        )
        descriptor.fetchLimit = 1
        return try context.fetch(descriptor).first
    }
}

File: android/app/src/main/java/com/mkflabs/inputpilot/data/DeviceRepository.kt
package com.mkflabs.inputpilot.data

import com.mkflabs.inputpilot.discovery.DiscoveredService
import com.mkflabs.inputpilot.discovery.SavedDeviceIndex
import com.mkflabs.inputpilot.network.Device
import com.mkflabs.inputpilot.network.DeviceApiClient
import com.mkflabs.inputpilot.network.DeviceEndpointResolver
import com.mkflabs.inputpilot.network.DeviceStatus
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.flow.Flow

sealed class DeviceRepositoryException(message: String) : Exception(message) {
    data object Unreachable : DeviceRepositoryException("Could not reach the device on the local network.")
    data object NotFound : DeviceRepositoryException("Device not found.")
    data class AlreadyExists(val displayName: String) :
        DeviceRepositoryException(SavedDeviceIndex.alreadyExistsMessage(displayName))
}

data class ProbedDevice(
    val candidate: DiscoveredService,
    val status: DeviceStatus,
    val baseUrl: String,
)

class DeviceRepository(
    private val dao: DeviceDao,
    private val api: DeviceApiClient,
) {
    fun observeDevices(): Flow<List<StoredDeviceEntity>> = dao.observeAll()

    suspend fun getAll(): List<StoredDeviceEntity> = dao.getAll()

    suspend fun refreshAll(devices: List<StoredDeviceEntity>): Set<String> = coroutineScope {
        val results =
            devices.map { device ->
                async {
                    val urls = DeviceEndpointResolver.endpointUrls(device.mdnsHost, device.staIp)
                    var status: DeviceStatus? = null
                    var fallback = device.mdnsHost
                    for (url in urls) {
                        runCatching { api.status(url, device.apiToken) }.onSuccess {
                            status = it
                            fallback = DeviceEndpointResolver.hostOf(url) ?: device.mdnsHost
                            return@async Triple(device.deviceId, status, fallback)
                        }
                    }
                    Triple(device.deviceId, null as DeviceStatus?, fallback)
                }
            }.awaitAll()

        val failed = mutableSetOf<String>()
        for ((deviceId, status, fallback) in results) {
            val device = devices.firstOrNull { it.deviceId == deviceId } ?: continue
            if (status != null) {
                dao.update(applyStatus(device, status, fallback))
            } else {
                failed.add(deviceId)
            }
        }
        failed
    }

    suspend fun addFromDiscovery(
        status: DeviceStatus,
        fallbackHost: String,
        displayName: String,
        token: String?,
    ): StoredDeviceEntity {
        val index = SavedDeviceIndex.fromDevices(dao.getAll())
        index.match(status, fallbackHost)?.let {
            throw DeviceRepositoryException.AlreadyExists(it.displayName)
        }
        val device =
            Device.fromStatus(status, fallbackHost).copy(
                displayName = displayName,
                apiToken = token,
            )
        val entity = StoredDeviceEntity.from(device)
        dao.upsert(entity)
        return dao.getById(entity.deviceId) ?: throw DeviceRepositoryException.NotFound
    }

    suspend fun probeByAddress(host: String, token: String?): ProbedDevice {
        val urls = DeviceEndpointResolver.endpointUrls(host, null)
        if (urls.isEmpty()) throw DeviceRepositoryException.Unreachable
        val index = SavedDeviceIndex.fromDevices(dao.getAll())
        var last: Throwable = DeviceRepositoryException.Unreachable
        for (url in urls) {
            try {
                val status = api.status(url, token)
                val fallback = DeviceEndpointResolver.hostOf(url) ?: host
                index.match(status, fallback)?.let {
                    throw DeviceRepositoryException.AlreadyExists(it.displayName)
                }
                val candidate =
                    DiscoveredService(
                        id = "manual-$fallback",
                        deviceId = status.deviceId,
                        name = status.name,
                        host = fallback,
                        port = 80,
                    )
                return ProbedDevice(candidate, status, url)
            } catch (e: DeviceRepositoryException.AlreadyExists) {
                throw e
            } catch (e: Throwable) {
                last = e
            }
        }
        throw last
    }

    suspend fun setJiggle(device: StoredDeviceEntity, enabled: Boolean) {
        val urls = DeviceEndpointResolver.endpointUrls(device.mdnsHost, device.staIp)
        if (urls.isEmpty()) throw DeviceRepositoryException.Unreachable
        var last: Throwable = DeviceRepositoryException.Unreachable
        for (url in urls) {
            try {
                api.setJiggle(url, enabled, device.apiToken)
                dao.update(device.copy(jiggleEnabled = enabled))
                return
            } catch (e: Throwable) {
                last = e
            }
        }
        throw last
    }

    suspend fun rename(deviceId: String, name: String) {
        val device = dao.getById(deviceId) ?: throw DeviceRepositoryException.NotFound
        dao.update(device.copy(displayName = name))
    }

    suspend fun delete(deviceId: String) {
        dao.deleteById(deviceId)
    }

    suspend fun updateApiToken(deviceId: String, token: String?) {
        val device = dao.getById(deviceId) ?: throw DeviceRepositoryException.NotFound
        val trimmed = token?.trim()?.takeIf { it.isNotEmpty() }
        dao.update(device.copy(apiToken = trimmed))
    }

    private fun applyStatus(
        stored: StoredDeviceEntity,
        status: DeviceStatus,
        fallbackHost: String,
    ): StoredDeviceEntity {
        val mdns =
            when {
                !status.mdns.isNullOrBlank() -> status.mdns
                stored.mdnsHost.isEmpty() -> fallbackHost
                else -> stored.mdnsHost
            }
        return stored.copy(
            mdnsHost = mdns,
            staIp = DeviceEndpointResolver.directAddress(status.staIp, fallbackHost) ?: stored.staIp,
            jiggleEnabled = status.jiggle,
            lastSeenEpochMs = System.currentTimeMillis(),
            firmwareVersion = status.version,
        )
    }
}
